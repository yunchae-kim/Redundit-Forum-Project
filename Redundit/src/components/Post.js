/* eslint-disable react/destructuring-assignment */
/* eslint-disable max-len */
import { Divider } from "@blueprintjs/core";
import { Link, useHistory } from "react-router-dom";
import { useEffect, useState } from "react";
import NavBar from "./NavBar";
import userIcon from "../resources/userIcon.png";
import { getPostById, deletePostById, flagPostById } from "../utility/axiosPost";
import LinkButton from "./LinkButton";
import { getHashtag } from "../utility/axiosHashtags";
import "../style/components/Post.css";
import CommentList from "./CommentList";
import httpClient from "../utility/auth.helper";
import { getMember } from "../utility/axiosMember";
import { addHide } from "../utility/axiosHide";
import { addNotification } from "../utility/axiosNotification";

const Post = ({ postId }) => {
  const [post, setPost] = useState({});
  const [fetched, setfetched] = useState(false);
  const [tags, setTags] = useState([]);
  const [userId, setUserId] = useState(0);
  const [member, setMember] = useState(null);
  const history = useHistory();

  useEffect(() => {
    const userObj = httpClient.getCurrentUser();
    if (userObj) {
      const { iduser } = userObj;
      if (iduser) {
        setUserId(iduser);
      }
    }
  }, [userId]);

  useEffect(() => {
    setfetched(true);
    getPostById(postId).then((postFetched) => {
      setPost(postFetched);
      getMember(postFetched.idgroup, userId).then((memberFetched) => {
        setMember(memberFetched);
      });
    });
  }, [fetched]);

  useEffect(() => {
    getHashtag(postId, "post").then((tagsFetched) => {
      setTags(tagsFetched);
    });
  }, []);

  const hashTagsList = tags.map((h) => (
    <Link to={{ pathname: `/hashtag/${h.hashtag}`, state: { hashtag: h.hashtag } }} key={h.hashtag}>
      #
      {h.hashtag}
      {" "}
    </Link>
  ));

  const deletePost = () => {
    deletePostById(postId).then(() => {
      if (member !== null && member.admin === 1) {
        addNotification(post.iduser, `Your post ${post.title} is deleted by admin.`).then(() => {
          history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
        });
      } else {
        history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
      }
    });
  };

  const editPost = () => {
    history.push({ pathname: `/editpost/post/${postId}`, state: { post, hashtags: tags } });
  };

  const flagPost = () => {
    flagPostById(postId, true);
  };

  const hidePost = () => {
    addHide(userId, postId, post.idgroup).then(() => {
      history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
      history.go();
    });
  };

  return (
    <div>
      <NavBar />
      <Link to={{ pathname: `/postanalytics/${postId}`, state: { postId } }}>Post Analytics</Link>
      <LinkButton url={`/group/${post.idgroup}`} text="Back to Group" groupId={post.idgroup} />
      <div className="PostPage">
        <div className="Post-header">
          <img className="Post-userIcon" src={userIcon} alt="" />
          <div className="Post-userName">
            {post.status === 1
              ? <Link to={{ pathname: `/user/${post.iduser}`, state: { userId: post.iduser } }}><span className="CW-chatentry-name">{post.username}</span></Link>
              : (
                <span className="CW-chatentry-name">
                  {post.username}
                  {" "}
                  (deactivated)
                </span>
              )}
          </div>
          <div className="Post-postDate">
            Posted:
            {" "}
            {post.date}
          </div>
        </div>
        <div className="Post-title">
          {post.title}
        </div>
        <div className="Post-hashtags">
          {hashTagsList}
        </div>
        <div className="Post-img-div">
          {post.image == null ? <></> : <img src={post.image} alt="" className="Post-img" />}
          {post.audio == null ? <></> : (
            <audio controls src={post.audio}>
              Your browser does not support the audio element.
              <track src="" kind="captions" srcLang="en" label="english_captions" />
            </audio>
          )}
          {post.video == null ? <></> : (
            <video controls src={post.video}>
              Your browser does not support the video element.
              <track src="" kind="captions" srcLang="en" label="english_captions" />
            </video>
          )}
        </div>
        <div className="Post-post">
          {post.content}
        </div>
        <hr />
        <div className="Post-fooder">
          {((userId !== 0 && userId !== post.iduser)) && <button type="submit" onClick={() => hidePost()}>Hide</button>}
          <Divider />
          {((member != null)) && <button type="submit" onClick={() => flagPost()}>Flag for Deletion</button>}
          <Divider />
          {((post.iduser === userId) || (member != null && member.admin === 1)) && <button type="submit" onClick={() => deletePost()}>Delete</button>}
          <Divider />
          {(post.iduser === userId) && <button type="submit" onClick={() => editPost()}>Edit</button>}
        </div>
      </div>
      <div className="PostPage-comments">
        <CommentList postId={postId} />
      </div>
    </div>
  );
};

export default Post;
