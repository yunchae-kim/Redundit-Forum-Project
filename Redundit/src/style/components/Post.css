.PostPage {
    width: 1100px;
    margin: 29px 100px 0 160px;
    padding: 17px 1px 21px 0;
    border-radius: 20px;
    background-color: #fff;
    font-family: HelveticaNeue;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: normal;
    letter-spacing: normal;
    text-align: left;
  }
  
  .Post-header {
    display: flex;
    flex-direction: row;
    margin-left: 30px;
    margin-right: 32px;
    padding-top: 0px;
    padding-bottom: 10px;
    color: black;
    font-size: 16px;
  }
  
  .Post-userIcon {
    border-radius: 30%;
    width: 20px;
    height: 20px;
    align-self: center;
    margin-right: 15px;
  }
  
  .Post-postDate {
    align-self: center;
  }
  .Post-userName {
    margin-right: 30px;
    align-self: center;
    font-weight: bold;
  }
  .Post-title {
    height: 28px;
    margin: 0px 167px 10px 30px;
    font-family: HelveticaNeue;
    font-size: 24px;
    font-weight: bold;
    color: #111;
  }
  
  .Post-hashtags {
    margin: 10px 318px 16px 30px;
    font-size: 16px;
    color: #465771;
  }
  
  .Post-post {
    width: 1038px;
    margin: 15px 36px 15px 30px;
    font-size: 20px;
    overflow: hidden;
    text-overflow: ellipsis;
    color: #111;
  }
  
  .Post-fooder {
    display: flex;
    width: 485px;
    height: 18px;
    margin: 15.5px 587px 0 30px;
    font-size: 16px;
    color: #188bd8;
  }
  
  .Post-viewPost{
    margin-top: 20px;
    margin-bottom: 20px;
    font-size: 20px;
    font-weight: bold;
    text-align: center;
    align-self: center;
  }
  
  .Post-viewPostText {
    color : #5da271;
  }
  
  .Post-img {
    height: 360px;
    width: 720px;
  }
  
  .Post-img-div {
    display: flex;
    justify-content: center;
    margin-top: 20px;
    margin-bottom: 20px;
  }
  
  .PostPage-comments {
    color: black;
    width: 1100px;
    margin: 24px 100px 0 160px;
    padding: 17px 1px 21px 0;
    border-radius: 20px;
    background-color: #fff;
    font-family: HelveticaNeue;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: normal;
    letter-spacing: normal;
    text-align: left;
  }
  
  .Post-comment {
    margin-left: 27px;
    margin-right: 27px;
    margin-bottom: 27px;
    font-size: 20px;
  }
  
  .Post-comment-modify {
    display: flex;
    color: #444444;
  }
  
  .Post-comment-modify a{
    font-size: 16px;
    color: #444444;
  }
  
  .Post-comment-name{
    font-size: 16px;
    font-weight:700;
    margin-bottom: 10px;
  }
  
  .Post-comment-time{
    font-weight: 300;
    color:#444444;
    margin-left: 17px;
  }
  
  .Post-comment-icon{
    border-radius: 30%;
    width: 20px;
    height: 20px;
    align-self: center;
    margin-right: 15px;
  }
  