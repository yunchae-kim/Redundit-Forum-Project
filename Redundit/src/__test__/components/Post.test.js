import {
  waitFor,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import Post from "../../components/Post";

test("test should Post match snapshot", async () => {
  const tree = renderer.create(
    <BrowserRouter>
      <Post postId={1} />
    </BrowserRouter>,
  ).toJSON();
  await waitFor(async () => {
    expect(tree).toMatchSnapshot();
  });
});
