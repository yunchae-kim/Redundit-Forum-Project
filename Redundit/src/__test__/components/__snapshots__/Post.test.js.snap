// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should Post match snapshot 1`] = `
<div>
  <div
    className="bp3-navbar bp3-fixed-top bp3-navbar navbar"
  >
    <div
      className="bp3-navbar-group bp3-align-left"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt=""
          className="navbar-logo"
          src="logo.svg"
        />
      </a>
    </div>
    <div
      className="bp3-navbar-group bp3-align-right"
    >
      <div>
        <a
          href="/login"
          onClick={[Function]}
        >
          <button
            className="bp3-button bp3-small bp3-intent-primary navbar-button"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            type="button"
          >
            <span
              className="bp3-button-text"
            >
              Log In
            </span>
          </button>
        </a>
        <a
          href="/signup"
          onClick={[Function]}
        >
          <button
            className="bp3-button bp3-small bp3-intent-primary navbar-button"
            onBlur={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            type="button"
          >
            <span
              className="bp3-button-text"
            >
              Sign Up
            </span>
          </button>
        </a>
      </div>
    </div>
  </div>
  <a
    href="/postanalytics/1"
    onClick={[Function]}
  >
    Post Analytics
  </a>
  <a
    href="/group/undefined"
    onClick={[Function]}
  >
    <button
      className="link-button"
      type="button"
    >
      Back to Group
    </button>
  </a>
  <div
    className="PostPage"
  >
    <div
      className="Post-header"
    >
      <img
        alt=""
        className="Post-userIcon"
        src="userIcon.png"
      />
      <div
        className="Post-userName"
      >
        <span
          className="CW-chatentry-name"
        >
           
          (deactivated)
        </span>
      </div>
      <div
        className="Post-postDate"
      >
        Posted:
         
      </div>
    </div>
    <div
      className="Post-title"
    />
    <div
      className="Post-hashtags"
    />
    <div
      className="Post-img-div"
    />
    <div
      className="Post-post"
    />
    <hr />
    <div
      className="Post-fooder"
    >
      <div
        className="bp3-divider"
      />
      <div
        className="bp3-divider"
      />
      <div
        className="bp3-divider"
      />
    </div>
  </div>
  <div
    className="PostPage-comments"
  >
    <div
      className="Cmnt-container"
    >
      <div
        className="Cmnt-"
      >
        <div
          id="Cmnt-input"
        >
          <div
            className="bp3-input-group bp3-large"
          >
            <input
              className="bp3-input"
              id="Cmnt-inputUser"
              onChange={[Function]}
              placeholder="Input comment"
              style={
                Object {
                  "paddingLeft": undefined,
                  "paddingRight": undefined,
                }
              }
              type="text"
            />
          </div>
        </div>
        <div
          id="NP-tagInput"
        >
          <ul
            id="NP-tags"
          />
          <input
            onKeyUp={[Function]}
            placeholder="Enter to add hashtags"
            type="text"
          />
        </div>
        <div
          className="NP-postBtnDiv"
        />
      </div>
      <div />
    </div>
  </div>
</div>
`;
